resources/2024-06-support-secretstores-env.md

```diff
@@ -0,0 +1,417 @@
+			}
+		} else {
+			m[key] = datamodel.EnvironmentVariable{
+				ValueFrom: &datamodel.EnvironmentVariableReference{
+					SecretRef: &datamodel.EnvironmentVariableSecretReference{
+						Source: to.String(val.ValueFrom.SecretRef.Source),
+						Key:    to.String(val.ValueFrom.SecretRef.Key),
+					},
+				},
+			}
+
+		}
+
+	}
+	return m, nil
+}
+
+// fromEnvDataModel: Converts from base datamodel to versioned datamodel
+func fromEnvDataModel(e map[string]datamodel.EnvironmentVariable) map[string]*EnvironmentVariable {
+	m := map[string]*EnvironmentVariable{}
+
+	for key, val := range e {
+		if val.Value != nil {
+			m[key] = &EnvironmentVariable{
+				Value: val.Value,
+			}
+		} else {
+			m[key] = &EnvironmentVariable{
+				ValueFrom: &EnvironmentVariableReference{
+					SecretRef: &EnvironmentVariableSecretReference{
+						Source: to.Ptr(val.ValueFrom.SecretRef.Source),
+						Key:    to.Ptr(val.ValueFrom.SecretRef.Key),
+					},
+				},
+			}
+
+		}
+	}
+
+	return m
+}
+
+```
+
+
+### API design 
+
+<!--
+Include if applicable – any design that changes our public REST API, CLI
+arguments/commands, or Go APIs for shared components should provide this
+section. Write N/A here if not applicable.
+- Describe the REST APIs in detail for new resource types or updates to
+  existing resource types. E.g. API Path and Sample request and response.
+- Describe new commands in the CLI or changes to existing CLI commands.
+- Describe the new or modified Go APIs for any shared components.
+-->
+Updates to the container typespec to allow for secret references in environment variables. This replaces the existing environment variable type of `map[string]string`  to allow for a secret reference.  
+
+**containers.tsp**
+```diff
+@doc("Definition of a container")
+model Container {
+  @doc("The registry and image to download and run in your container")
+  image: string;
+
+  @doc("The pull policy for the container image")
+  imagePullPolicy?: ImagePullPolicy;
+
+  @doc("environment")
+  env?: Record<EnvironmentVariable>;
+
+  @doc("container ports")
+  ports?: Record<ContainerPortProperties>;
+
+  @doc("readiness probe properties")
+  readinessProbe?: HealthProbeProperties;
+
+  @doc("liveness probe properties")
+  livenessProbe?: HealthProbeProperties;
+
+  @doc("container volumes")
+  volumes?: Record<Volume>;
+
+  @doc("Entrypoint array. Overrides the container image's ENTRYPOINT")
+  command?: string[];
+
+  @doc("Arguments to the entrypoint. Overrides the container image's CMD")
+  args?: string[];
+
+  @doc("Working directory for the container")
+  workingDir?: string;
+}
+
+@doc("Envinronment variables type")
+model EnvironmentVariable {
+
+  @doc("The value of the environment variable")
+  value?: string;
+
+  @doc("The reference to the variable")
+  valueFrom?: EnvironmentVariableReference;
+}
+
+@doc("The reference to the variable")
+model EnvironmentVariableReference {
+  @doc("The secret reference")
+  secretRef: SecretReference;
+}
```

Comment:
Note: we anticipate adding (non-secret) references to configuration, or other sources in the future. That's why this additional level is needed.

Reply:
Resolved in offline conversations